### DISCUSSION TITLE
Collision Resolution

### DISCUSSION THREAD

**Discussion by asherhe** [bepu/bepuphysics2#245] [Q&A] [answered]
Hello!

Just out of curiosity, how does the collision resolution system work?

**Comment by RossNordby** [comment, marked as answer]
At a high level:

1. Broad phase finds collision pairs that *might* be touching by comparing bounding boxes. A bounding volume hierarchy is used to accelerate testing.
2. Pairs with overlapping bounds are handed off to the `CollisionBatcher`. It waits until there are enough pairs of a given type (for example, box vs cylinder, or sphere vs capsule) and then dispatches a batched test over all the pairs in the batch. Those batched tests can be internally [vectorized](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) for better performance.
3. The collision batcher outputs contacts for each pair. Contacts contain information like the offset from the involved bodies to the contact point, the surface normal at the contact point, and the penetration depth at the contact point. For a given pair of bodies, these contact points will be processed into a set of 1-4 contacts that approximate the contact surface. Sometimes, contacts are created between bodies that are not actually touching and have *negative* depth but [may touch soon](https://github.com/bepu/bepuphysics2/blob/master/Documentation/ContinuousCollisionDetection.md).
4. Contact *constraints* are created from those sets of contacts. In bepuphysics2, constraints attempt to modify the velocities of the involved bodies to prevent (or resolve) penetration. Solving for the velocities necessary to mitigate penetration for one pair may cause penetrating velocities for *another* pair, though- consider a big pile of bodies. To converge to a solution that works for the entire pile at once, the pairwise constraints are solved repeatedly. Over time, this converges to a decent global solution. 
5. This iteration can either take the form of repeating the velocity constraint solving process multiple times in one frame- termed velocity iterations in bepuphysics2- or you can take smaller solver timesteps ([substeps](https://github.com/bepu/bepuphysics2/blob/master/Documentation/Substepping.md)), usually with fewer velocity iterations per substep. For simulations with lots of very stiff/rigid constraints, substepping can improve stability enormously within the same computational budget.
6. While constraints operate at the velocity level, they can still correct position error (for contacts, penetration between bodies) by including a corrective term in the target velocity that they solve for. In bepuphysics2, this behavior is exposed through spring settings; you can specify the oscillation frequency and damping ratio for all position constraints. In fact, bounciness in bepuphysics2 is just configuring contact constraints to be springier. If you go into super slow mo and watched the substeps, you'd see objects briefly go into penetration and then get pushed out, approximating the deformation that results in bounciness in the real world.
